src/Propulsion.Kafka/Consumers.fs

```diff
@@ -283,62 +283,150 @@ module Core =
+                keyValueToStreamEvents : KeyValuePair<string, string> -> Propulsion.Streams.StreamEvent<_> seq,
+                handle : StreamName * Streams.StreamSpan<_> -> Async<'Res>, maxDop,
+                stats : Streams.Scheduling.StreamSchedulerStats<OkResult<'Res>, FailResult>,
+                /// Prevent batches being consolidated prior to scheduling in order to maximize granularity of consumer offset updates
+                ?maximizeOffsetWriting,
+                ?pipelineStatsInterval, ?maxSubmissionsPerPartition, ?pumpInterval, ?logExternalState, ?idleDelay, ?maxBatches) =
+            StreamsConsumer.Start<KeyValuePair<string, string>, 'Res>(
+                log, config, Bindings.mapConsumeResult, keyValueToStreamEvents, handle, maxDop, stats,
+                ?pipelineStatsInterval = pipelineStatsInterval,
+                ?maxSubmissionsPerPartition = maxSubmissionsPerPartition,
+                ?pumpInterval = pumpInterval,
+                ?logExternalState = logExternalState,
+                ?idleDelay = idleDelay,
+                ?maxBatches = maxBatches,
+                ?maximizeOffsetWriting = maximizeOffsetWriting)
+
+    // Maps a (potentially `null`) message key to a valid {Category}-{StreamId} StreamName for routing and/or propagation through StreamsProjector
+    let parseMessageKey defaultCategory = function
+        | null -> FsCodec.StreamName.create defaultCategory ""
+        | key -> Propulsion.Streams.StreamName.parseWithDefaultCategory defaultCategory key
+
+/// StreamsConsumer buffers and deduplicates messages from a contiguous stream with each message bearing an `index`.
+/// Where the messages we consume don't have such characteristics, we need to maintain a fake `index` by keeping an int per stream in a dictionary
+type StreamNameSequenceGenerator() =
+
+    // we synthesize a monotonically increasing index to render the deduplication facility inert
+    let indices = System.Collections.Generic.Dictionary()
+
+    /// Generates an index for the specified StreamName
+    member __.GenerateIndex(streamName : StreamName) =
+        let streamName = FsCodec.StreamName.toString streamName
+        match indices.TryGetValue streamName with
+        | true, v -> let x = v + 1 in indices.[streamName] <- x; int64 x
+        | false, _ -> let x = 0 in indices.[streamName] <- x; int64 x
+
+    /// Provides a generic mapping from a ConsumeResult to a <c>StreamName</c> and <c>ITimelineEvent</c>
+    member __.ConsumeResultToStreamEvent
+        (   toStreamName : ConsumeResult<_, _> -> StreamName,
+            toTimelineEvent : ConsumeResult<_, _> * int64 -> ITimelineEvent<_>)
+        : ConsumeResult<_, _> -> Propulsion.Streams.StreamEvent<byte[]> seq =
+        fun consumeResult ->
+            let sn = toStreamName consumeResult
+            let e = toTimelineEvent (consumeResult, __.GenerateIndex sn)
+            Seq.singleton { stream = sn; event = e }
```

Comment:
Why wrap this in a Seq?

Reply:
that's the signature demanded by the `StreamsConsumer.Start` signature - its key to
a) being able to drop messages that are not relevant
b) being able to collect multiple items encoded in a single event (e.g.. RenderedSpans format)

For these specific overloads I do appreciate it looks incongruous

Arguably, `toStreamName`'s result could be changed to be an `option` - where it returning None could then imply Seq.empty